AsyncDisplayKit/ASCollectionView.mm

```diff
@@ -1143,57 +1136,32 @@ - (void)nodeDidRelayout:(ASCellNode *)node sizeChanged:(BOOL)sizeChanged
     return;
   }
   
-  BOOL queued = (_queuedNodeSizeInvalidationContext != nil);
-  if (!queued) {
-    _queuedNodeSizeInvalidationContext = [[_ASCollectionViewNodeSizeInvalidationContext alloc] init];
-    
-    __weak __typeof__(self) weakSelf = self;
-    dispatch_async(dispatch_get_main_queue(), ^{
-      __typeof__(self) strongSelf = weakSelf;
-      if (strongSelf) {
-        [strongSelf requeryNodeSizes];
-      }
-    });
+  NSIndexPath *indexPath = [self indexPathForNode:node];
+  if (indexPath == nil) {
+    return;
   }
+
+  [_layoutFacilitator collectionViewWillEditCellsAtIndexPaths:@[ indexPath ] batched:NO];
   
-  [_queuedNodeSizeInvalidationContext.invalidatedNodes addObject:node];
+  ASCollectionViewInvalidationAction previousAction = _nextLayoutInvalidationAction;
+  ASCollectionViewInvalidationAction action = previousAction;
+  if (previousAction == ASCollectionViewInvalidationActionNone) {
+    [self setNeedsLayout];
+    action = ASCollectionViewInvalidationActionInvalidateWithAnimation;
+  }
 
-  // Check if this node or one of its subnodes can be animated.
-  // If the context is already non-animated, don't bother checking this node.
-  if (_queuedNodeSizeInvalidationContext.shouldAnimate) {
+  // If we think we're going to animate, check if this node will prevent it.
+  if (action == ASCollectionViewInvalidationActionInvalidateWithAnimation) {
     BOOL (^shouldNotAnimateBlock)(ASDisplayNode *) = ^BOOL(ASDisplayNode * _Nonnull node) {
```

Comment:
It's silly, but this block will cause measurable retain / release overhead in profiles.  Not an issue now, but we should earmark this for integration into ASEnvironment state propagation.  There are nice mechanisms available to be able to answer this kind of question by checking state on one node.  The state propagation occurs whenever a node with the non-default value is added as a subnode; if one is removed as a subnode, only in the case that the one removed has a non-default value does the tree need to be re-queried to check if the state can be returned to default, and since removal of subnodes is less common this turns out to be a win.